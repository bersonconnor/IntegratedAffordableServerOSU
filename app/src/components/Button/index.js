import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import "./scss/button.scss";

const Button = ({
  buttonLabelText,
  containerClassName,
  handleClickByParent,
  kind,
  ...props
}) => {
  const handleClick = event => {
    handleClickByParent(event);
  };

  return (
    <div
      className={classNames("button__container", {
        [containerClassName]: containerClassName
      })}
    >
      <button
        className={classNames("button", `button--${kind}`)}
        onClick={handleClick}
        {...props}
      >
        {buttonLabelText}
      </button>
    </div>
  );
};

Button.propTypes = {
  /*
   * Displays text label for button
   */
  buttonLabelText: PropTypes.string.isRequired,
  /*
   * Contains classes to style the container surrounding button
   */
  containerClassName: PropTypes.string,
  /**
   * The style of button to create. Primary buttons correspond with primary actions
   * on a page, secondary and tertiary are, likewise, corresponding to secondary and
   * tertiary actions on a page
   */
  kind: PropTypes.oneOf(["primary", "secondary", "tertiary"]),
  /**
   * Logic that handles change event if state is kept
   * in parent node
   */
  handleClickByParent: PropTypes.func
};

Button.defaultProps = {
  containerClassName: undefined,
  handleClickByParent: () => {},
  kind: "primary"
};

export default Button;
