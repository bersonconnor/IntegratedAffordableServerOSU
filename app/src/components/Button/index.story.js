import React, { Fragment, useState } from "react";
import { storiesOf } from "@storybook/react";
import Button from "./index";

const SimpleComponent = () => {
  const [isVisible, setIsVisible] = useState(false);

  return (
    <Fragment>
      {isVisible && <p>Hello World!</p>}
      <Button
        buttonLabelText="Say Hello"
        handleClickByParent={() => {
          setIsVisible(!isVisible);
        }}
      />
    </Fragment>
  );
};

storiesOf("Button", module)
  .addDecorator(storyFun => {
    return <div style={{ padding: "1rem" }}>{storyFun()}</div>;
  })
  .add("primary", () => <Button buttonLabelText="Button" kind="primary" />)
  .add("secondary", () => <Button buttonLabelText="Button" kind="secondary" />)
  .add("tertiary", () => <Button buttonLabelText="Button" kind="tertiary" />)
  .add("simple component", () => <SimpleComponent />);
