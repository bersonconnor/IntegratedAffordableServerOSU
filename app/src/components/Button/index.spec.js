import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Button from "./";

const testButtonLabelTest = "TestLabel";

const testButton = (
  <Button buttonLabelText={testButtonLabelTest} kind="primary" type="button" />
);

describe("Button", () => {
  describe("Snapshots", () => {
    it("should match snapshot for default Button", () => {
      expect(renderer.create(testButton).toJSON()).toMatchSnapshot();
    });
  });

  describe("label text", () => {
    it("should render string passed in", () => {
      const wrapper = shallow(testButton);
      expect(wrapper.find("button")).toHaveText(testButtonLabelTest);
    });
  });

  describe("onClick functionality", () => {
    it("should call parent function when button is clicked", () => {
      const mockHandleClick = jest.fn(() => {});
      const wrapper = shallow(
        <Button buttonLabelText="test" handleClickByParent={mockHandleClick} />
      );
      wrapper.find("button").simulate("click");
      expect(mockHandleClick.mock.calls.length).toBe(1);
    });
  });
});
