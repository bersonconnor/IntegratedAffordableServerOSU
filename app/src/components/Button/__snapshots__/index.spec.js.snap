// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshots should match snapshot for default Button 1`] = `
<div
  className="button__container"
>
  <button
    className="button button--primary"
    onClick={[Function]}
    type="button"
  >
    TestLabel
  </button>
</div>
`;
